### PR TITLE
kustomize edit add replacement

### LINKED ISSUE #4679: kustomize edit add replacement
**Is your feature request related to a problem? Please describe.**
We got some scripts which generates a replacement file , and would like to automatically add the replacement file to the `kustomization.yaml` file. _(I patch in a init container, and then we want that init container to have the same image as a main container. I know, it's hacky, but I am testing some PoCs)_ Similarly to how we can add a patch with `kustomize edit add patch --path` or a label with `kustomize edit add label`.

I've recently started working with replacements, so maybe I am missing something in the documentation or the intentions with it, but I can't see any support for this feature and I would consider it really useful.

### PR DESCRIPTION
**Pull request by charles-chenzz**
fix #4679

### CODE REVIEW

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [inline kustomize/commands/edit/add/addreplacement.go:17]: I've local tested and it didn't work fine. I suspect this field need more setup
image


can you take a look when you have a chance? @annasong20

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:17]: image

nvm, seems like I figure out. will tweak the code when I got more time

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:30]: suggestion: it'd be nice for UX to expand the example here showing the whole `kustomize edit ...` command, akin to what the examples shown in `kustomize edit add --help` do.

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:16]: question: are there any specific benefits from using the type `ReplacementField` here, since only `path` seems to be allowed?

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:74]: suggestion: combine the two and only return the error message with the information of which `path` failed.
```suggestion
			return fmt.Errorf("replacement for path %q already in %s file", r.Path, konfig.DefaultKustomizationFileName())
```

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement_test.go:1]: suggestion: combine these tests as a single function with a test case table, since they all use the same setup.

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement_test.go:39]: suggestion: since this is verifying that the command will fail if a replacement for the same path is already present, perhaps the kustomization file could be initialized with that content?

File: kustomize/commands/edit/add/all.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/all.go:52]: Good call adding the example here! 🎉

**Review by stormqueen1990 — COMMENTED** [review]
Left some suggestions!

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:16]: I ref ``edit add patch`` and use ``ReplacementField`` in case we need more fields in ``edit add replacement``
does this make sense?

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:74]: good catch! update comes later

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:31]: this is follow what ``kustomize edit add patch`` done before

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement_test.go:1]: https://github.com/kubernetes-sigs/kustomize/pull/5444#discussion_r1400439332

same like this

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement_test.go:39]: not really needed. because I add another cmd execute down below so that if you don't write anything in kustomization file and still be fine

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [inline kustomize/commands/edit/add/addreplacement.go:58]: the lint check is interesting.. it's same as the ``edit add patch`` and it still failed the check

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement.go:17]: Yes, that's fair.

File: kustomize/go.sum
Comment (stormqueen1990) [inline kustomize/go.sum:74]: It seems something went wrong during rebase here.

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement.go:58]: This might be because `edit add patch` was created before the linters were enabled.

File: kustomize/go.sum
Comment (charles-chenzz) [reply kustomize/go.sum:74]: uh.. yes. I didn't check the rebase file after rebase.. should remove the ``<<<HEAD``, ``====``, and ``>>>4b1e...``

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:58]: seems like this is the reason. then I need to follow the lint hint to wrap it I guess

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement_test.go:36]: Is there a more meaningful assertion that could be done here?
I'm concerned this test might pass by happenstance without verifying that the result was what we expected to happen.

A test similar to this would be more reliable:
```go
kf, err := kustfile.NewKustomizationFile(fSys)
require.NoError(t, err)

kustomization, err := kf.Read()
require.NoError(t, err)

for _, replacement := range kustomization.Replacements {
    require.Equal(t, expectedPath, replacement.Path)
}
```

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:74]: this is a trick or hack. while I was developing this feature, I didn't reset here at first and then I build and run. then found out that I can't get the result we want. then I was think maybe is b/c the replacement fields then I reset it. surprisingly found out that everything works fine with this reset

Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement.go:74]: Can this be avoided if we remove the initialization on lines 23-24?

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go:74]: will remove the initialization on lines 23-24 and let CI to check, update comes later today

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: CI passed, haven't tested in local build

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:24]: Could we remove this commented block entirely?
```suggestion
```

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:73]: This also looks like it's not needed any longer.
```suggestion
```

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: removed

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: removed

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement_test.go]: update as suggested

File: go.work.sum
Comment (stormqueen1990) [inline go.work.sum:1]: It doesn't look like your code changes should have updated this (I don't see any new modules). Do you think this can be reverted for this PR?

File: kustomize/commands/edit/add/addreplacement_test.go
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement_test.go]: Thanks for updating this!

**Review by stormqueen1990 — COMMENTED** [review]
Good work here! 🎉

This is mostly LGTM, I just see one lint failure. I think we could also revert the `go.work.sum` change, since your code is not adding new modules. What do you think?

File: go.work.sum
Comment (charles-chenzz) [reply go.work.sum:1]: the update is resolved the need-rebase label

**Review by stormqueen1990 — COMMENTED** [review]
/lgtm

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:27]: From what I've seen, the inline string case is not supported by the code in this PR, so we should update the long description to match what the command does:
```suggestion
		Long: `this command will add an item to replacement field in the kustomization file.
The item must be a file path.
```

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: updated. sorry for late reply

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [inline kustomize/commands/edit/add/addreplacement.go:27]: Minor word leftover
```suggestion
The item must be a file path.
```

File: kustomize/commands/edit/add/addreplacement.go
Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement.go]: "either" was still leftover 😅 I left a comment with that suggestion.

Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: sorry about that 😓 updated

File: kustomize/commands/edit/add/addreplacement.go
Comment (charles-chenzz) [reply kustomize/commands/edit/add/addreplacement.go]: sorry, updated it

Comment (stormqueen1990) [reply kustomize/commands/edit/add/addreplacement.go]: No worries 🙂

**Review by stormqueen1990 — COMMENTED** [review]
/lgtm

**Comment by charles-chenzz** [comment]
would you mind approve @stormqueen1990 ? seems like you're approver of command(except build)

File: kustomize/commands/edit/add/addreplacement.go
Comment (koba1t) [inline kustomize/commands/edit/add/addreplacement.go:71]: Hi @charles-chenzz 
Could you add a check to file exists of replacement file?

I think other `edit add` command is implemented that.

**Review by koba1t — CHANGES_REQUESTED** [review]

**Review by varshaprasad96 — APPROVED** [review]
Went over the implementation and looks good to me. Lgtm after Yugo's review is addressed. 

/hold - Adding hold so that it doesn't auto merge till Yugo's lgtm is in place.

**Pull request closed without merging**
